include/RubyEngine.hpp

```diff
@@ -0,0 +1,56 @@
+#ifndef __RUBYENGINE_HPP__
+#define __RUBYENGINE_HPP__
+
+#include <string>
+#include <mutex>
+#include <mesos/mesos.hpp>
+
+const std::string PARAM_SCRIPT_PATH = "script_path";
+
+namespace criteo {
+  namespace mesos {
+
+    class RubyEngine
+    {
+      int state = 0;
+
+    public:
+      RubyEngine(const std::string& name);
+
+      ~RubyEngine();
+
+      bool load_script(const std::string& scriptname);
+
+      bool is_callback_defined(const std::string& symbol);
+
+      std::string handle_exception();
+    };
+
+    // This class is a singleton,
+    // to avoid multiple initializations
+    // from multiple modules
+    //
+    // /!\ First set of mesos module paramater wins!
+    //
+    class UniqueRubyEngine : public RubyEngine
+    {
+      bool started = false;
+      std::mutex m_mutex;
+
+    public:
+      static UniqueRubyEngine& getInstance();
+      void start(const ::mesos::Parameters& parameters);
```

Comment:
FAR: there is a way to completely remove this method from the API since what it does was not very clear to me when only reading the API.
I would have created a Singleton at higher level called RubyModule that has 2 methods: Isolator* getIsolatorInstance(const Parameters&) and Hook* getHookInstance(const Parameters&) that would have been called only in the factory method of the hook and isolator and have executed the code currently in "start" method. That way, you have no ambiguity on the usage of "start" since it does not exist anymore.

Reply:
This is not equivalent in terms of features: we may want several hooks and/or isolators, but it would not be possible if the isolator and hook were a singleton

Reply:
You can implement it in ruby and keep the cpp simple.